src/inspector_profiler.cc

```diff
@@ -503,6 +507,217 @@ static void StopCoverage(const FunctionCallbackInfo<Value>& args) {
+    watchdog->Stop();
+  }
+}
+
+void ProfileWatchdog::Run(void* arg) {
+  ProfileWatchdog* wd = static_cast<ProfileWatchdog*>(arg);
+  uv_run(&wd->loop_, UV_RUN_DEFAULT);
+}
+
+void ProfileWatchdog::Timer(uv_timer_t* timer) {
+  ProfileWatchdog* w = ContainerOf(&ProfileWatchdog::timer_, timer);
+  Environment* env = w->env();
+  if (!env || env->is_stopping()) {
+    return;
+  }
+  if (w->state() == ProfileWatchdogState::kClosed) {
+    return;
+  }
+  env->RequestInterrupt([watchdog = std::move(w)](Environment* env) {
+    if (env->is_stopping() ||
+        watchdog->state() == ProfileWatchdogState::kClosed) {
+      return;
+    }
+    if (watchdog->state() == ProfileWatchdogState::kClosing) {
+      delete watchdog;
+      return;
+    }
+    if (watchdog->TimeoutHandler()) {
+      watchdog->AddTask(
+          [watchdog = std::move(watchdog)]() { watchdog->SetTimeout(); });
+    }
+  });
+}
+
+template <typename Fn>
+void ProfileWatchdog::AddTask(Fn&& cb, CallbackFlags::Flags flags) {
+  auto callback = tasks_.CreateCallback(std::move(cb), flags);
+  {
+    Mutex::ScopedLock lock(task_mutex_);
+    tasks_.Push(std::move(callback));
+  }
+  uv_async_send(&async_);
+}
+
+void ProfileWatchdog::HandleTasks() {
+  while (tasks_.size() > 0) {
+    CallbackQueue<void> queue;
+    {
+      Mutex::ScopedLock lock(task_mutex_);
+      queue.ConcatMove(std::move(tasks_));
+    }
+    while (auto head = queue.Shift()) head->Call();
+  }
+}
+
+void ProfileWatchdog::SetTimeout() {
+  int rc = uv_timer_start(&timer_, &ProfileWatchdog::Timer, interval_, 0);
+  CHECK_EQ(0, rc);
+}
+
+Local<FunctionTemplate> ProfileWatchdog::GetConstructorTemplate(
+    Environment* env) {
+  Local<FunctionTemplate> tmpl = env->profile_watchdog_ctor_template();
+  if (tmpl.IsEmpty()) {
+    Isolate* isolate = env->isolate();
+    tmpl = NewFunctionTemplate(isolate, nullptr);
+    tmpl->SetClassName(
+        FIXED_ONE_BYTE_STRING(env->isolate(), "ProfileWatchdog"));
+    SetProtoMethod(isolate, tmpl, "start", ProfileWatchdog::Start);
+    SetProtoMethod(isolate, tmpl, "stop", ProfileWatchdog::Stop);
+    env->set_profile_watchdog_ctor_template(tmpl);
+  }
+  return tmpl;
+}
+
+ProfileWatchdog::ProfileWatchdog(Environment* env, Local<Object> object)
+    : BaseObject(env, object) {
+  int rc;
+
+  rc = uv_loop_init(&loop_);
+  CHECK_EQ(0, rc);
+
+  rc = uv_async_init(&loop_, &async_, [](uv_async_t* task_async) {
+    ProfileWatchdog* w = ContainerOf(&ProfileWatchdog::async_, task_async);
+    w->HandleTasks();
+  });
+
+  CHECK_EQ(0, rc);
+
+  rc = uv_timer_init(&loop_, &timer_);
+  CHECK_EQ(0, rc);
+}
+
+ProfileWatchdog::~ProfileWatchdog() {
+  set_state(ProfileWatchdogState::kClosed);
+  AddTask([self = std::move(this)]() {
+    uv_close(reinterpret_cast<uv_handle_t*>(&self->timer_), nullptr);
+    uv_close(reinterpret_cast<uv_handle_t*>(&self->async_), nullptr);
+    uv_stop(&self->loop_);
+  });
+  uv_thread_join(&thread_);
+  uv_run(&loop_, UV_RUN_DEFAULT);
```

Comment:
You should check the return value of uv_thread_join(), it can fail.

Calling uv_run() from multiple threads is undefined behavior, even if you do it sequentially. Cleanup should happen on the other thread.

(Also applies to uv_timer_start(). Once you kickstart the event loop, it belongs to the thread it's running on.)

Reply:
Thanks, i will update the code(`uv_timer_start` is called in worker thread currently).